client.go

```diff
@@ -0,0 +1,194 @@
+func (c *Client) CreateConcept(concept Concept, task string) error {
+	if concept.PrefLabel == "" {
+		return errors.New("input concept should have prefLaber defined")
+	}
+
+	if concept.SchemaObject == "" {
+		return errors.New("input concept should have schema defined")
+	}
+
+	if concept.Type == "" {
+		return errors.New("input concept should have type defined")
+	}
+
+	// Construct the request url. It looks like smartlogicURL?path=task:MyModel:Mytask/skos:Concept/rdf:instance.
+	reqURL := c.baseAPIURL
+	path := fmt.Sprintf("task:%s:%s/skos:Concept/rdf:instance", c.model, task)
+	// We don't want to encode the path param here.
+	reqURL.RawQuery = fmt.Sprintf("path=%s", path)
+
+	// Construct the request body, we really on the custom marshalling of the concept object.
+	body, err := json.Marshal(concept)
+	if err != nil {
+		return fmt.Errorf("failed json encoding concept: %w", err)
+	}
+	resp, err := c.makeAuthorizedRequest("POST", reqURL.String(), bytes.NewBuffer(body))
+	if err != nil {
+		return fmt.Errorf("failed creating new concept: %v", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
+		return fmt.Errorf("failed creating new concept, returned status %v", resp.StatusCode)
+	}
+
+	return nil
+}
+
+func (c *Client) AddConceptMetadataField(conceptID, fieldName, fieldValue, task string) error {
+	// Construct the request url. It looks like smartlogicURL?path=task:MyModel:Mytask/doubleEncodedConcept.
+	reqURL := c.baseAPIURL
+
+	conceptURI := ConceptURIPrefix + "/" + conceptID
+	// Smartlogic API requires the conceptURI that is part of the path query param to be escaped twice and inside < >.
+	encodedConceptURI := url.QueryEscape(url.QueryEscape(fmt.Sprintf("<%s>", conceptURI)))
+
+	path := fmt.Sprintf("task:%s:%s/%s", c.model, task, encodedConceptURI)
+
+	// We don't want to encode the path param here.
+	reqURL.RawQuery = fmt.Sprintf("path=%s", path)
+
+	// Construct the request body.
+	fieldURI := MetadataFieldPrefix + "/" + fieldName
+	bodyMap := map[string]string{
+		"@id":    conceptURI,
+		fieldURI: fieldValue,
+	}
+	body, err := json.Marshal(bodyMap)
+	if err != nil {
+		return fmt.Errorf("failed encoding metadata body: %w", err)
+	}
+
+	resp, err := c.makeAuthorizedRequest("POST", reqURL.String(), bytes.NewBuffer(body))
+	if err != nil {
+		return fmt.Errorf("failed adding metadata to concept %s: %v", conceptID, err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
+		return fmt.Errorf("failed adding metadata to concept %s, returned status %v", conceptID, resp.StatusCode)
+	}
+
+	return nil
+}
+
+func (c *Client) makeAuthorizedRequest(method, url string, body io.Reader) (*http.Response, error) {
+	for accessFailures := 0; accessFailures < MaxAccessFailures; accessFailures++ {
+		req, err := http.NewRequest(method, url, body)
```

Comment:
If the intention is to provide a base code to be reused in our services when communicating with Smartlogic, I would like to see a `*WithContext` interface for better flow control. (`CreateConceptWithContext`, `AddConceptMetadataFieldWithContext`)
```suggestion
func (c *Client) makeAuthorizedRequest(ctx context.Context, method, url string, body io.Reader) (*http.Response, error) {
	for accessFailures := 0; accessFailures < MaxAccessFailures; accessFailures++ {
		req, err := http.NewRequestWithContext(ctx, method, url, body)
```

Reply:
👍